src/MQTT433gateway.cpp

```diff
@@ -197,65 +160,95 @@ void rfRawCallback(const uint16_t *pulses, int length) {
+        mqttClient->sendRaw(data);
+      }
     }
   }
 }
 
-void reconnect() {
-  beatLED.on();
-  // Loop until we're reconnected
-  while (!mqtt.connected()) {
-    Debug(F("Attempting MQTT connection..."));
-    // Attempt to connect
-    if (mqtt.connect(mainTopic.c_str(), mqttUser, mqttPassword,
-                     mainTopic.c_str(), 0, true, "offline")) {
-      DebugLn(F("connected"));
-      mqtt.publish(mainTopic.c_str(), "online", true);
-      mqtt.subscribe((mainTopic + F("/set/+")).c_str());
-      mqtt.subscribe((mainTopic + F("/ota/+")).c_str());
-      mqtt.subscribe((mainTopic + F("/send/+")).c_str());
-      mqtt.subscribe((globalTopic + F("/send/+")).c_str());
-    } else {
-      Debug(F("failed, rc="));
-      Debug(mqtt.state());
-      DebugLn(F(" try again in 5 seconds"));
-      // Wait 5 seconds before retrying
-      beatLED.off();
-      delay(500);
-      beatLED.on();
-      delay(4500);
-    }
+void reconnectMqtt(const Settings &) {
+  if (mqttClient != nullptr) {
+    delete mqttClient;
+    mqttClient = nullptr;
   }
-  beatLED.off();
+
+  mqttClient = new MqttClient(settings, wifi);
+  mqttClient->begin();
+
+  mqttClient->onSet(F("log"),
+                    [](const String &payload) { logMode = payload[0] == '1'; });
+  mqttClient->onSet(F("raw"),
+                    [](const String &payload) { rawMode = payload[0] == '1'; });
+  mqttClient->onSet(F("protocols"), [](const String &payload) {
+    settings.updateProtocols(payload);
+  });
+
+  mqttClient->onRfData(transmitt);
+  mqttClient->onOta(handleOta);
+}
+
+// ToDO: default init should go to settings!
+void defaultSettings() {
+  settings.deviceName = String(F("rfESP_")) + String(ESP.getChipId(), HEX);
+
+  settings.mqttReceiveTopic = settings.deviceName + F("/recv/");
+  settings.mqttLogTopic = settings.deviceName + F("/log/");
+  settings.mqttRawRopic = settings.deviceName + F("/raw/");
+
+  settings.mqttSendTopic = settings.deviceName + F("/send/");
+  settings.mqttConfigTopic = settings.deviceName + F("/set/");
+  settings.mqttOtaTopic = settings.deviceName + F("/ota/");
+
+  settings.mqttBroker = F(myMQTT_BROCKER);
+  settings.mqttBrokerPort = 1883;
+  settings.mqttUser = static_cast<const char *>(myMQTT_USERNAME);
+  settings.mqttPassword = static_cast<const char *>(myMQTT_PASSWORD);
+  settings.mqttRetain = true;
+
+  settings.rfEchoMessages = false;
+  settings.rfProtocols = F("[]");
+
+  settings.otaPassword = F(myOTA_PASSWD);
 }
 
 void setup() {
   Serial.begin(115200, SERIAL_8N1, SERIAL_TX_ONLY);
 
-  if (!connectWifi(ssid, password, []() { beatLED.loop(); })) {
+  if (!connectWifi(mySSID, myWIFIPASSWD, []() { beatLED.loop(); })) {
     DebugLn(F("Try connectiing again after reboot"));
     ESP.restart();
   }
 
-  mqtt.setServer(mqttBroker, 1883);
-  mqtt.setCallback(mqttCallback);
+  settings.onChange(MQTT, reconnectMqtt);
+  settings.onChange(
+      RF_ECHO, [](const Settings &s) { rf.setEchoEnabled(s.rfEchoMessages); });
```

Comment:
looks a little bit ugly, but efficient.

Reply:
Debug will hopefully be replaced in general. I haven't made up my mind completely so far how to do it. I thought of a proxy object that know log levels an where you can register streams for levels. Like serial, a websockets connection, syslog or whatever. But maybe there is also a good lib out there for this purpose on these devices. I'll have a look later.

Reply:
Debugging here, at this comment? But anyhow, a Stream like proxy object that can redirect to multiple Streams sounds good. Log levels, at least: debug, log and info. Yes, and best would be a existing lib.

Reply:
I used an mobile app for commenting and it confused the comments somhow. This was meant as a comment to your comment https://github.com/puuu/MQTT433gateway/pull/21#discussion_r155920803

Sorry.

Reply:
I could not find an existing lib that fit my needs. So I created https://github.com/janLo/ArduinoSimpleLogging.